### PR TITLE
Uniffi Places Interrupt methods

### PR DESCRIPTION
**Pull request by skhamis**
### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt
Comment (mhammond) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt:105]: I don't think this is correct - it should remain using the old technique while we still have that duality of both handle and uniffi connections. We will need *something* to kill the uniffi side of the world, but that might just be able to be a `this.writeConn.close()` method.

(To be clear, this probably isn't going to hurt because we will kill the old handle stuff before merging to main, but it still makes sense to me to keep things clear)

File: components/places/src/ffi.rs
Comment (mhammond) [inline components/places/src/ffi.rs:136]: as above, I think just removing this from this PR is fine. We probably do need a new ticket for "remove all old handle cruft and ensure we have appropriate cleanup/close functions for the new world"

File: components/places/src/places.udl
Comment (mhammond) [inline components/places/src/places.udl:32]: In theory it *should* be used by ios so they can cancel a long running sync on shutdown, but yeah, for now, this is fine (maybe update the comment to also say the sync manager will typically arrange to interrupt syncs or something)

**Review by mhammond — COMMENTED** [review]
This looks roughly correct! It's a shame things are broken already, but I don't see how this makes anything worse and I think it is on the right path.

**Review by tarikeshaq — COMMENTED** [review]
I think this might be our first use of having an interface reference another interface 🤩 Looks good! Dropped a couple of thoughts

Something that maybe we can follow-up on later after we are done uniffication:

We usually don't have interfaces referencing other interfaces across the ffi because I imagine it usually happens all in Rust, is it possible to move the setup for the interrupt handler into Rust? 

So the [`PlacesConnection`](https://github.com/mozilla/application-services/blob/9067babc820f04d87b6d18504ae6b816980067c5/components/places/src/ffi.rs#L136) struct would hold an `interrupt_handler` field as opposed to the kotlin class. Then the consumer can call the `interrupt` method on the `PlacesConnection` class, and it would internally deal it.

That said, I'm not sure what the reasoning was to setup the API like this in the first place..

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt
Comment (skhamis) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt:105]: hmmm this is an interesting point. So it actually makes sense to simply not uniffi these connection-level handle problems and mainly just focus on being able to call interrupt on the new connections.

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt
Comment (skhamis) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt:1102]: 🤦 that's what I get for putting this PR up way too early in the draft process haha. Noted!

**Comment by skhamis** [comment]
The swiftlint is unrelated to this PR but everything else is passing, and no identified changes needed in either consumer!

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt
Comment (mhammond) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt:219]: I have a very vague preference for just `InterruptHandle` here, because it's (a) more than just sql and (b) kinda leaks an implementation detail out (eg, ultimately "tabs" should expose an interrupt handle too, but it doesn't even use sql!)

OTOH, I'm really just saying I hate the name in the rust code, but the UDL having the same name as the rust code does make some sense.

So ultimately I'll leave that up to you.

File: components/places/src/ffi.rs
Comment (mhammond) [inline components/places/src/ffi.rs:151]: The semantics here also smell wrong. Given the arcs etc, I think the right approach would be for this to be called something like `get_interrupt_handle()` and the struct itself gets a `Weak<SqlInterruptHandle>` or something (the idea being there's exactly 1 interrupt handle per connection and getting multiple "new" ones per connection makes no sense) - but it doesn't make sense to fix this now, so please just add a comment referencing 1684 too.

(And if you get bored, starting to tackle 1684 makes some sense to me - we might decide to land it after we land this effort, but there's no good reason to not get started now.)

**Review by mhammond — APPROVED** [review]
LGTM - it's a shame the shape of this api gets more broken than it already was, but I think it's fine to land with this initially to help keep our blast radius down 💣

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt
Comment (skhamis) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt:219]: hmm unfortunately the `interface SqlInterruptHandle` in UDL is in reference to the rust side. So I believe the three options we have are:
1. Rename SqlInterrupt handle in the support component (used quite a bit though)
2. Typealias in our kotlin/swift wrappers to keep the name (though still leaks a tiny bit)
3. Keep same

I have no strong preference either way but would probably lean to leaving as-is if only to keep things simple for now until the interrupt rework takes place.

File: components/places/src/ffi.rs
Comment (skhamis) [reply components/places/src/ffi.rs:151]: I completely agree it feels weird to get a new handle everytime and will add proper comments,

**Pull request merged**
